crates/next-api/src/client_references.rs

```diff
@@ -0,0 +1,70 @@
+use std::collections::HashMap;
+
+use anyhow::Result;
+use next_core::{
+    self, next_client_reference::EcmascriptClientReferenceModule,
+    next_server_component::server_component_module::NextServerComponentModule,
+};
+use serde::{Deserialize, Serialize};
+use turbo_tasks::{
+    debug::ValueDebugFormat, trace::TraceRawVcs, ResolvedVc, TryFlatJoinIterExt, Vc,
+};
+use turbopack::css::CssModuleAsset;
+use turbopack_core::module::Module;
+
+use crate::module_graph::SingleModuleGraph;
+
+#[derive(Clone, Serialize, Deserialize, Eq, PartialEq, TraceRawVcs, ValueDebugFormat)]
+pub enum ClientReferenceMapType {
+    EcmascriptClientReference {
+        module: ResolvedVc<EcmascriptClientReferenceModule>,
+        ssr_module: ResolvedVc<Box<dyn Module>>,
+    },
+    CssClientReference(ResolvedVc<CssModuleAsset>),
+    ServerComponent(ResolvedVc<NextServerComponentModule>),
+}
+
+#[turbo_tasks::value(transparent)]
+pub struct ClientReferencesSet(HashMap<ResolvedVc<Box<dyn Module>>, ClientReferenceMapType>);
+
+#[turbo_tasks::function]
+pub async fn map_client_references(
+    graph: Vc<SingleModuleGraph>,
+) -> Result<Vc<ClientReferencesSet>> {
```

Comment:
To improve incremental builds this could get a `Vc<ModulesSet>` instead of the full graph. Since a `ModulesSet` is `PartialEq` we could cache that step for some changes of the graph.

Reply:
I'll try that in a future PR.
(Effectively this set would be the `nodes: Vec<Node<N, Ix>>` that is currently stored in the petgraph.)